inputgraph/loader.go

```diff
@@ -393,94 +423,82 @@ func (l *loader) loadBlock(ctx context.Context, b spec.Block) error {
 	return nil
 }
 
-func (l *loader) hashIfStatement(s spec.IfStatement) {
-	l.hasher.HashString("IF")
-	l.hasher.HashJSONMarshalled(s.Expression)
-	l.hasher.HashBool(s.ExecMode)
-	l.hasher.HashInt(len(s.IfBody))
-	l.hasher.HashInt(len(s.ElseIf))
-	if s.ElseBody != nil {
-		l.hasher.HashInt(len(*s.ElseBody))
-	}
-}
-
-func (l *loader) hashElseIf(e spec.ElseIf) {
-	l.hasher.HashString("ELSE IF")
-	l.hasher.HashJSONMarshalled(e.Expression)
-	l.hasher.HashBool(e.ExecMode)
-	l.hasher.HashInt(len(e.Body))
-}
-
-func (l *loader) hashWaitStatement(w spec.WaitStatement) {
-	l.hasher.HashString("WAIT")
-	l.hasher.HashInt(len(w.Body))
-	l.hasher.HashJSONMarshalled(w.Args)
-}
-
-func (l *loader) hashVersion(v spec.Version) {
-	l.hasher.HashString("VERSION")
-	l.hasher.HashJSONMarshalled(v.Args)
-}
+func (l *loader) forTarget(ctx context.Context, target domain.Target, args []string) (*loader, error) {
+	fullTargetName := target.String()
 
-func (l *loader) hashCommand(c spec.Command) {
-	l.hasher.HashString(c.Name)
-	l.hasher.HashJSONMarshalled(c.Args)
-	l.hasher.HashBool(c.ExecMode)
-}
+	visited := copyVisited(l.visited)
+	visited[fullTargetName] = struct{}{}
 
-func (l *loader) hashFor(f spec.ForStatement) {
-	l.hasher.HashString("FOR")
-	l.hasher.HashJSONMarshalled(f.Args)
-}
+	flagArgs, err := variables.ParseFlagArgs(args)
+	if err != nil {
+		return nil, err
+	}
 
-func copyVisited(m map[string]struct{}) map[string]struct{} {
-	m2 := map[string]struct{}{}
-	for k := range m {
-		m2[k] = struct{}{}
+	// TODO: confirm these are set correctly.
+	overriding, err := variables.ParseCommandLineArgs(flagArgs)
```

Comment:
Does this look correct? I'm looking to pass build args to the new loader instance. It seems to work.

Reply:
I should clarify. What I'm unsure of is whether using `OverridingVars` is the expected way to set build args when using `variables.Collection`.

This seems to be the case based on some uses elsewhere:
https://github.com/earthly/earthly/blob/mh%2Fauto-skip-arg-passing/earthfile2llb/converter.go#L137
https://github.com/earthly/earthly/blob/mh%2Fauto-skip-arg-passing/variables/collection.go#L93

Reply:
I take back my "I think so!" comment, it's rather murky :sweat_smile: 

when the converter ends up in `buildTarget`, it calls prepBuildTarget: https://github.com/earthly/earthly/blob/410c690e4df488287225c713da16d817a3488932/earthfile2llb/converter.go#L1781

which then calls some similar variables functions:

https://github.com/earthly/earthly/blob/410c690e4df488287225c713da16d817a3488932/earthfile2llb/converter.go#L1740C2-L1748

note that it depends on if `--pass-args` is given, or if we're calling a target in the same Earthfile, or a different one.

which then gets stored under `OverridingVars`, which then gets referenced under https://github.com/earthly/earthly/blob/410c690e4df488287225c713da16d817a3488932/earthfile2llb/converter.go#L1791-L1831

there's a lot of logic here -- maybe we can move it out of the `earthfile2llb` package, and into a common util which both the inputgraph and earthfile2llb packages could call?

Reply:
Thanks. I ran into this code as well. I think it may be hard to extract given that the `Converter` code is [bound](https://github.com/earthly/earthly/blob/mh%2Fauto-skip-arg-passing/earthfile2llb/converter.go#L1800) to the `MultiTarget`. But I think the basic idea of propagating them is the same. I'll also have to account for `--pass-args` & `--build-args` I think. Anyhow, sounds like I'm on the right track!